elixir/apps/domain/lib/domain/auth/adapters/openid_connect.ex

```diff
@@ -110,17 +114,24 @@ defmodule Domain.Auth.Adapters.OpenIDConnect do
             nil
         end
 
-      Identity.Query.by_id(identity.id)
+      # we try to use `email` if it's exposed, but fall back to OpenID Connect `sub` claim
```

Comment:
This is a good call, however I wonder if we should just have a hard requirement to require the `email` claim to be there. We could always relax this requirement later on. If we can assume that each user has an email it could simplify some other features in the product that will rely on outbound emails.

There have been a couple cases where this hasn't been the case:

- Most commonly on Azure: https://github.com/firezone/firezone/issues/1684 but the workaround/fix is fairly straightforward there
- https://github.com/firezone/firezone/issues/1572 Authgear doesn't seem to be widely used and doesn't seem to implement the spec appropriately here, so we can avoid building for this edge case

-

Reply:
I was suggesting perhaps we make the product decision to keep the `email` claim as a hard requirement in Firezone because it will make a number of things easier to reason about:

- We can assume every user can receive an outbound emails for things like notifications, reports
- Admin signs up with magic link, connects an IdP, then migrates his/her account to a connected IdP
- Admin A signs up, makes user B an Admin, then user B demotes Admin A. Do we allow user B to be the sole admin of this tenant with no email?
- Billing reminders, invoices, etc

The two examples I gave in my previous comment were cases where our current hard requirement for `email` may have caused an issue, but in the end were resolved without issue.

We could also prompt for an email when encountering an of the scenarios above, but that might get messy?

Reply:
Ah, now I get it. 

I'm not a big fan of using email claims at all, orgs have a policy for email reuse but even then there is a grace period, (eg. you will be able to reuse it in 2 years), while in Firezone we don't know about it and will eventually log a user in into somebody else account. It's even worse if the company has no email reuse policy.

So maybe instead we can drop `email` claim entirely, use `sub` which is part of the spec, unique, and always there, and just use `email` to fill out the profile for notifications (and warn if there is none to manually enter it)?

Reply:
Email reuse policies are an interesting edge case I hadn't thought of.

Yeah, we can stick to the spec as best we can, and then require emails when needed if they're missing. Another edge case I thought was for support -- we'll need an email to communicate with the user if he/she opens a support ticket. 🙃

Reply:
@AndrewDryga, I agree with your thoughts on email reuse and we should use the `sub` claim for identification on the Firezone side of things.  That said, I can also understand where @jamilbk is coming from since we may need to be able to communicate with these users outside of the Firezone application.

I may be thinking about this wrong, but I think we could do both.  We could use the `sub` claim as the Firezone identifier and also require that the `email` claim be populated too if we wanted.  We shouldn't need to worry about logging someone in to an old account at that point, because the IDP would only ever give the `sub` id for the currently active owner of the email right?

If what I've outlined above holds true, businesses may like the ability to control what email is being set for any given Firezone account they "own", rather than allowing the user to fill in whatever email they want in their profile.  e.g. If I'm an employee of `Company A`, I probably shouldn't be allowed to set my Firezone profile email to `user@gmail.com`, but rather be forced to use my company email `user@companya.com`

I'm sure there's probably something I'm not thinking through, but figured I'd throw this out and see what everyone thought

Reply:
That's a good point @bmanifold. I think for the purposes of this PR we should probably clarify that OIDC claims are only used for sign in, so it does make sense to use `sub` to find the user and match on it for that.

The tricky part is the SCIM integration. For sync, _if_ we find a user's email matching a particular `sub` we can populate it, but if not, we can't really do anything at that point. We probably still want to allow the user to authenticate via OIDC, but we're at the mercy of the sync'd attributes from SCIM to populate the email.

I think @AndrewDryga's suggestion to allow them to populate the email was good, but perhaps we require the admin to set this up successfully via the SCIM connector instead of letting the user do it.

For reference, I did [some research](https://github.com/firezone/product/issues/461) on which providers support SCIM over OIDC and which don't (or require SAML):

| Provider | SCIM support |
| --- | --- |
| Auth0 | Supported by [professional services only](https://auth0.com/docs/authenticate/protocols/scim). GA expected Q4 '23 |
| Google | Only supported for [apps in the directory](https://apps.google.com/supportwidget/articlehome?hl=en&article_url=https%3A%2F%2Fsupport.google.com%2Fa%2Fanswer%2F7681608%3Fhl%3Den&assistant_id=generic-unu&product_context=7681608&product_name=UnuFlow&trigger_context=a). Needs a custom adapter otherwise |
| Okta | [SAML only](https://help.okta.com/en-us/Content/Topics/Apps/Apps_App_Integration_Wizard_SCIM.htm) |
| Azure AD | Supported for [both OIDC and SAML](https://learn.microsoft.com/en-us/azure/active-directory/app-provisioning/use-scim-to-provision-users-and-groups) |
| JumpCloud | [Supported for SAML](https://jumpcloud.com/support/provision-and-manage-users-and-groups-in-apps-using-custom-scim-identity-management-integration) |
| OneLogin | [Supported for SAML](https://onelogin.service-now.com/support?id=kb_article&sys_id=1c8d87f2dbcfad1057fac24505961922) |
| AWS IAM Identity Center | [Supported for SAML](https://docs.aws.amazon.com/singlesignon/latest/userguide/scim-profile-saml.html) |
| Authentik | Supported for [both OIDC and SAML](https://goauthentik.io/docs/providers/scim/) |
| Keycloak | Not supported |
| Ory Stack | [Not supported](https://github.com/ory/kratos/issues/1028) |